File: backend/ingest/wildfires.py
import requests
import logging
from typing import List
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)

# NASA FIRMS VIIRS active fire data (no-auth version uses CSV)
FIRMS_URL = "https://firms.modaps.eosdis.nasa.gov/data/active_fire/noaa-20-viirs-c2/csv"

def fetch_wildfires(days: int = 1) -> List[dict]:
    """Fetch active wildfire hotspots from NASA FIRMS."""
    # Use the 24h global dataset (no API key required for this endpoint)
    url = f"{FIRMS_URL}/J1_VIIRS_C2_Global_{days}d.csv"

    try:
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        lines = response.text.strip().split('\n')
        if len(lines) < 2:
            return []

        headers = [h.strip() for h in lines[0].split(',')]
        fires = []
        for line in lines[1:1001]:  # Limit to 1000
            values = line.split(',')
            if len(values) < len(headers):
                continue
            row = dict(zip(headers, values))
            try:
                fires.append({
                    "id": f"{row.get('latitude','0')}_{row.get('longitude','0')}_{row.get('acq_date','')}_{row.get('acq_time','')}",
                    "lat": float(row.get('latitude', 0)),
                    "lon": float(row.get('longitude', 0)),
                    "brightness": float(row.get('bright_ti4', 0)),
                    "frp": float(row.get('frp', 0)),
                    "acqDate": row.get('acq_date', ''),
                    "acqTime": row.get('acq_time', ''),
                    "confidence": row.get('confidence', 'n'),
                    "satellite": row.get('satellite', 'VIIRS'),
                })
            except (ValueError, KeyError):
                continue
        return fires
    except Exception as e:
        logger.error(f"Failed to fetch wildfires: {e}")
        return []

File: api/wildfires.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';

// Proxies NASA FIRMS CSV data and returns JSON
// Fetches VIIRS 24h global data
// Parses CSV and returns array of fire hotspots

export default async function handler(req: VercelRequest, res: VercelResponse) {
  const url = 'https://firms.modaps.eosdis.nasa.gov/data/active_fire/noaa-20-viirs-c2/csv/J1_VIIRS_C2_Global_24h.csv';

  try {
    const response = await fetch(url, {
      headers: { 'User-Agent': 'WeatherIntelligence/1.0' }
    });
    const text = await response.text();
    const lines = text.trim().split('\n');

    if (lines.length < 2) {
      res.json({ fires: [] });
      return;
    }

    const headers = lines[0].split(',').map(h => h.trim());
    const fires = [];

    for (const line of lines.slice(1, 1001)) {
      const values = line.split(',');
      if (values.length < headers.length) continue;
      const row: Record<string, string> = {};
      headers.forEach((h, i) => { row[h] = values[i]?.trim() || ''; });

      try {
        fires.push({
          id: `${row.latitude}_${row.longitude}_${row.acq_date}_${row.acq_time}`,
          lat: parseFloat(row.latitude),
          lon: parseFloat(row.longitude),
          brightness: parseFloat(row.bright_ti4 || '0'),
          frp: parseFloat(row.frp || '0'),
          acqDate: row.acq_date,
          acqTime: row.acq_time,
          confidence: row.confidence,
          satellite: row.satellite || 'VIIRS',
        });
      } catch { continue; }
    }

    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Cache-Control', 's-maxage=3600, stale-while-revalidate=600');
    res.json({ fires });
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch wildfire data', fires: [] });
  }
}
